### PR TITLE
feat(Core/Authserver): TOTP rewrite

### LINKED ISSUE #5397: Suggestion: securely store TOTP secrets [$25]
We probably should look at porting this to more securely store the TOTP secrets:
https://github.com/TrinityCore/TrinityCore/commit/4211645834c467a03c60248e80818d3607be9ea7

Currently the TOTP secrets are stored in clear text in the auth database.

<bountysource-plugin>

---
There is a **[$25 open bounty](https://www.bountysource.com/issues/97973633-suggestion-securely-store-totp-secrets?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Base and original commit WarheadCore/Warhead@4211645
- Cherrypick https://github.com/WarheadCore/WarheadBand/commit/0adab3454d139c9ac73f95ee40da510af22af36b

## Issues Addressed:
- Closes #5397
- Part #3451

## Tests Performed:
- Login and enter world

## How to Test the Changes:
- CREATE BACKUP DB TABLE `account`
- Apply all db changes
- Test default enter game (world)
- If you know how to use TOTP, try :D (I don't know how to use it)

## Target Branch(es):
- [x] Master

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
Need #5518 before

**Comment by Kitzunu** [comment]
What is totp?

File: data/sql/updates/pending_db_auth/rev_1620079951672711500.sql [hunk 1]
```diff
@@ -0,0 +1,297 @@
+--
+
+START TRANSACTION; -- we're messing with the accounts table here, let's play it safe
+
+INSERT INTO `version_db_auth` (`sql_rev`) VALUES ('1620079951672711500');
+
+DROP TABLE IF EXISTS `secret_digest`;
+CREATE TABLE `secret_digest` (
+  `id` int(10) unsigned not null,
+  `digest` varchar(100) not null,
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_auth/rev_1620079951672711500.sql:10]:
```suggestion
  `id` INT UNSIGNED NOT NULL,
  `digest` VARCHAR(100) NOT NULL,
```

File: data/sql/updates/pending_db_auth/rev_1620079951672711500.sql [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+-- BASE32 CONVERSION ENDS HERE                         --
+-- (this is the other banner i promised you, so you    --
+--  can stop skipping the unnecessarily complex stuff) --
+-- =================================================== --
+
+ALTER TABLE `account` ADD COLUMN `totp_secret` VARBINARY(128) DEFAULT NULL AFTER `session_key`;
+UPDATE `account` a LEFT JOIN `_temp_totp_conversion` c ON a.`token_key`=c.`original_key` SET a.`totp_secret`=c.`totp_secret`;
+ALTER TABLE `account` DROP COLUMN `token_key`;
+
+COMMIT; -- safety gloves off
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_auth/rev_1620079951672711500.sql:297]:
```suggestion
COMMIT; -- safety gloves off

```

Comment (Kitzunu) [reply data/sql/updates/pending_db_auth/rev_1620079951672711500.sql:297]: newline

**Comment by Winfidonarleyan** [comment]
> What is totp?

https://ru.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm

**Comment by Winfidonarleyan** [comment]
Ready to test

**Comment by locus313** [comment]
Thanks, @Winfidonarleyan ill test this later today.

**Comment by locus313** [comment]
This looks good. The conversion of existing secrets worked without any problems. removing and setting new 2fa worked great.

**Comment by locus313** [comment]
yeah, think this can be merged

**Review by Si1ker — APPROVED** [review]

**Comment by Si1ker** [comment]
#5518 is already merged as WInfi pointed out, let's go with this one. 🐱

**Comment by Winfidonarleyan** [comment]
Let's go merge

**Pull request merged**